### ISSUE TITLE
Compile error with -DMPQC_R12 and HAVE_PSI3 (but no -DMPQC_NEW_FEATURES)

### ISSUE THREAD

**Issue by mbanck** [ValeevGroup/mpqc#41]
If I enable psi3 support (but not MPQC_NEW_FEATURES), I get the following build failure:

```
[ 95%] Building CXX object src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/psicc_pt2r12.cc.o
cd "/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/obj-x86_64-linux-gnu/src/lib/chemistry/qc/psi" && /usr/bin/c++   -DHAVE_CONFIG_H -DMPQC_R12 -g -O2 -std=c++11  -I"/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/obj-x86_64-linux-gnu/src/lib" -I"/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib" -I"/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/include" -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -I/usr/include/eigen3 -I/usr/include/libint2    -o CMakeFiles/psi.dir/psicc_pt2r12.cc.o -c "/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc_pt2r12.cc"
/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc_pt2r12.cc: In member function ‘void sc::PsiCC_PT2R12::compute_ept2r12()’:
/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc_pt2r12.cc:180:42: error: no matching function for call to ‘sc::PsiCC_PT2R12::T2_distarray4(const sc::SpinCase2&)’
     T2[s] = this->T2_distarray4(spincase2);
                                          ^
/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc_pt2r12.cc:180:42: note: candidate is:
In file included from /home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc_pt2r12.h:33:0,
                 from /home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc_pt2r12.cc:36:
/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc.h:63:23: note: sc::Ref<sc::DistArray4> sc::PsiCC::T2_distarray4(sc::SpinCase2, const string&)
       Ref<DistArray4> T2_distarray4(SpinCase2 spin, const std::string& L);
                       ^
/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psicc.h:63:23: note:   candidate expects 2 arguments, 1 provided
src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/build.make:80: recipe for target 'src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/psicc_pt2r12.cc.o' failed
make[3]: *** [src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/psicc_pt2r12.cc.o] Error 1
make[3]: Leaving directory '/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/obj-x86_64-linux-gnu'
CMakeFiles/Makefile2:2985: recipe for target 'src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/all' failed
make[2]: *** [src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/all] Error 2
make[2]: Leaving directory '/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/obj-x86_64-linux-gnu'
```

**Comment by mbanck** [comment]
Same build error with -DMPQC_NEW_FEATURES

**Comment by mbanck** [comment]
That fixed that compile problem, but I got another further down the road now:

```
[ 95%] Building CXX object src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/psiexenv.cc.o
cd "/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/obj-x86_64-linux-gnu/src/lib/chemistry/qc/psi" && /usr/bin/c++   -DHAVE_CONFIG_H -DMPQC_R12 -g -O2 -std=c++11  -I"/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/obj-x86_64-linux-gnu/src/lib" -I"/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib" -I"/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/include" -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -I/usr/include/eigen3 -I/usr/include/libint2    -o CMakeFiles/psi.dir/psiexenv.cc.o -c "/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psiexenv.cc"
/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psiexenv.cc: In member function ‘void sc::PsiExEnv::run_psiclean(bool)’:
/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/src/lib/chemistry/qc/psi/psiexenv.cc:337:11: error: ‘class psi::PSIO’ has no member named ‘purge’
     psio_.purge(fullclean);
           ^
src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/build.make:126: recipe for target 'src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/psiexenv.cc.o' failed
make[3]: *** [src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/psiexenv.cc.o] Error 1
make[3]: Leaving directory '/home/mbanck/debian/debichem/packages/mpqc3-0.0~git20150912/obj-x86_64-linux-gnu'
CMakeFiles/Makefile2:2985: recipe for target 'src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/all' failed
make[2]: *** [src/lib/chemistry/qc/psi/CMakeFiles/psi.dir/all] Error 2
```

**Comment by mbanck** [comment]
So this is due to MPQC shipping an unreleased (?, see http://sourceforge.net/projects/psicode/files/psi/) 3.5ish snapshot version of PSI3, which indeed has a purge() method for libpsio - I tried with the publicly available code (i.e., 3.4) in Debian, which does not.

@evaleev - I assume that psi3 version is not really 3.5, as confirmed by configure.ac identifying it as 3.4... If this was a stable release of PSI3 I'd be happy to upload that to Debian in case it gets put up somewhere more official like SourceForge or Github. But it looks like it is in an unfinished (for PSI3, not MPQC3) state. I'd hate to duplicate all that code all over again, so a compile-time option would be appreciated as well. I also had a quick stab at backporting the purge() method to 3.4 but it looks non-trivial to me.

Alternatively, one would have to live with the PSI3 files not being cleaned up for now.

**Comment by evaleev** [comment]
@mbanck this is fixed in fbec9cfae83b976e7c05176f76300f8059e68f28

This is indeed a patched version of Psi3. Patches did not make it into the tree since at that point I did not control/have access to Psi3 repo and our effort was no longer coordinated with the Psi team. This is moot as MPQC might be the only code that depends on Psi3 (I understand that Psi3 is still a part of official distributions ... I would just hate to spend any precious time on this: MPQC4 is a much higher priority).

**Comment by mbanck** [comment]
@evaleev that's fair enough, I wasn't aware that there was no access to the psicode SF project anymore for you.

I cleaned up the hacks I did earlier to make the PSI interface work with the released psicode-3.4.0 by adding CMake support in order to detect the version and added support for running psiclean instead of the newer purge() method in libpsio, the patch seems to work fine, check0 passes for the PSi checks, see e.g. https://buildd.debian.org/status/fetch.php?pkg=mpqc3&arch=amd64&ver=0.0~git20160216-3&stamp=1455870087

```
psiccsdpt2f12_h2opsiccsd_tf12ccpvdzccpvdzf12cabsccpvtzric2v_ansatz=diag,ebc=false_: ok ok E: 9
psiccsdpt2f12_h2opsiccsd_tf12ccpvdzccpvdzf12cabsccpvtzric2v_ansatz=ijkl,ebc=false_: ok ok E: 9
psipt2f12_casscfc11ccpvdzaugccpvdzc2v: ok ok E:99
psipt2f12_mrcic10ccpvdzaugccpvdzc2v: ok ok E:99
```

The patch is here: http://anonscm.debian.org/viewvc/debichem/unstable/mpqc3/debian/patches/check_for_psio_purge.patch?revision=6209&view=markup
The remaining wart is that I ignore non-zero exit status from the final psiclean as I saw frequent SystemExceptions otherwise I couldn't really figure out and it does not matter at that point, really.

**Comment by evaleev** [comment]
@mbanck actually I have access to SF, just checked. I have compared psi-3.4.0 and the version included in MPQC and clearly the latter uses unreleased code from what was to become psi 3.5 . That release never happened and I was not involved at that point.

Unfortunately trying to create a release based on the Psi code in MPQC would be a lot of work (releases in Psi 3 involved removing code not ready for release semiautomatically ... hence the superproject structure of Psi 4, with plugins etc.) ... given the resource constraints this is a nonstarter.

psio::purge was (as I recall) implemented to fix psiclean from doing bad things in nonstandard situations (using psi from mpqc is definitely one of them) where it was blowing away files it was not supposed to. replacing it with psiclean will work in many cases (after all, we were happy with "dumb" psiclean until after 3.4.0) but not all.

as always: thanks for your help.
